File: Arrays & Hashing/0260. Single Number III/0260.py
class Solution:
    def singleNumber(self, nums: List[int]) -> List[int]:
        n = len(nums)
        prefix = 0
        for num in nums:
            prefix ^= num
        
        diff_bit = 1 # 00001
        while not (diff_bit & prefix):
            diff_bit <<= 1
        
        a = b = 0

        for num in nums:
            if num & diff_bit:
                a ^= num
            else:
                b ^= num
        
        return [a, b]

File: Arrays & Hashing/0260. Single Number III/0260.cpp
#pragma GCC optimize("O3","Oz","unroll-loops","inline-functions","fast-math","march=native","lto","omit-frame-pointer")
class Solution {
public:
    vector<int> singleNumber(vector<int>& nums) {
        ios_base::sync_with_stdio(0);
        cin.tie(0);
        cout.tie(0);

        int32_t prefix = 0;
        for (auto& num : nums) {
            prefix ^= num;
        }
        
        int32_t diffBit = 1;
        while (!(diffBit & prefix)) {
            diffBit <<= 1;
        }

        int a = 0;
        int b = 0;
        for (auto& num : nums) {
            if (num & diffBit) {
                a ^= num;
            } else {
                b ^= num;
            }
        }

        return {a, b};
    }
};